File: PY/linked_list.py
# linked_list by Daniel Barbis

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

    def __str__(self):
        return str(self.data)

class Storage:
    def __init__(self):
        self.head = None
        self.items = 0

    # add item to the front of the list and increment the items counter
    def push(self, data):
        newNode = Node(data)
        if not self.head:
            self.head = newNode
        else:
            newNode.next = self.head
            self.head = newNode
        self.items += 1

    # remove the item at the front of the list while returning the data it contained and decrement the items counter
    def pop(self):
        if not self.head:
            return None
        data = self.head.data
        self.head = self.head.next
        self.items -= 1
        return data

    # return the data of the current head of the list
    def peek(self):
        return self.head.data

    # function to check if a linked list is empty or not
    def isempty(self):
        if not self.items:
            return True
        return False

    # function to swap data between nodes at index i and j
    def swap(self, i, j):
        if self.items <= i or self.items <= j and i is not j:
            return False

        node1 = None
        node2 = None

        temp = self.head

        k = 0
        while temp != None: 
            if k == i:
                node1 = temp
            if k == j:
                node2 = temp

            k += 1
            temp = temp.next


        temp = node1.data
        node1.data = node2.data
        node2.data = temp

        print(node1.data + " : " + node2.data)

        return True

File: CPP/linked_list.cpp
#include <iostream>
#include <typeinfo>

// this could / should probably be a struct instead since we're defaulting to public anyways
// difference between class and struct is none other than the default visibility
// class being private as default and struct being public
class Node
{
public:
    int data;
    Node* next = nullptr;

    // default constructor / destructor
    Node() = default;
    ~Node() = default;

    // constructor with data
    Node(int data) { this->data = data; }
};

class Storage
{
public:
    Node* head = nullptr;
    int index = 0;

    // default constructor / destructor
    Storage() = default;
    ~Storage() = default;

    // add item to the front of the list
    void push(int data)
    {
        Node* newNode = new Node(data);
        if (!head)
        {
            this->head = newNode;
        }
        else
        {
            newNode->next = this->head;
            this->head = newNode;
        }
        index++;
    }

    // remove item from the front of the list and return the value of said item
    bool pop(int &data)
    {
        if (!this->head)
            return false;

        data = this->head->data;
        auto temp = this->head->next;
        delete this->head;
        this->head = temp;

        index--;

        return data;
    }

    // return the data of the current head of the list
    bool peek(int &data)
    {
        if (!this->head)
            return false;
 
        data = this->head->data;
        return data;
    }

    // checks if the linked list if empty or not by checking if we have a head object in the list
    // if a head is present the list is not empty, a list cannot have any items without the head
    bool isEmpty()
    {
        if (!this->head)
            return true;
        return false;
    }

    // Not necessary for the assignment but this function will loop the linked list and print the data
    bool printList()
    {
        if (!this->head)
            return false;

        auto temp = this->head;

        while (temp != nullptr)
        {
            std::cout << "Data: " << temp->data << "\n";

            temp = temp->next;
        }

        return true;
    }

    // This function will find the nodes of which index we want to swap
    // It then swaps the data between the nodes and returns true
    bool swap(int i, int j)
    {
        if (this->index <= i || this->index <= j && i != j)
            return false;

        Node* node1 = nullptr;
        Node* node2 = nullptr;

        auto temp = this->head;

        int k = 0;
        while (temp != nullptr)
        {
            if (k == i)
                node1 = temp;
            if (k == j)
                node2 = temp;

            k++;
            temp = temp->next;
        }

        auto tempNode = *node1;
        node1->data = node2->data;
        node2->data = tempNode.data;

        delete temp;

        return true;
    }
};
